### PR TITLE
Support at modifier

### PR DESCRIPTION
**Pull request by yeya24**
Fixes https://github.com/thanos-community/promql-engine/issues/72

This pr adds support for at modifier. 
The current code is a bit messy tbh but I don't know if there are better ways to improve it. We don't support sub query and matrix selector without function so some code have not been added to make this feature complete.

### CODE REVIEW

File: physicalplan/step_invariant/step_invariant.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+func (u *stepInvariantOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	in, err := u.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil || len(in) == 0 || len(in[0].Samples) == 0 {
+		return nil, nil
+	}
+	if !u.duplicateResults {
```
Comment (fpetkovski) [inline physicalplan/step_invariant/step_invariant.go:88]: What does this flag indicate?

Comment (yeya24) [reply physicalplan/step_invariant/step_invariant.go:88]: Used in https://github.com/prometheus/prometheus/blob/main/promql/engine.go#L1675.
If not matrix or subquery then we will duplicate the values with steps.

**Comment by fpetkovski** [comment]
Thanks, I'll take a look at this more closely tomorrow. I want to make sure we are not simply importing tech debt from the old engine.

File: physicalplan/plan.go [hunk 2]
```diff
@@ -1,48 +1,69 @@
...
 )
 
 const stepsBatch = 10
 
 // New creates new physical query execution plan for a given query expression.
 func New(expr parser.Expr, storage storage.Queryable, mint, maxt time.Time, step, lookbackDelta time.Duration) (model.VectorOperator, error) {
-	return newCancellableOperator(expr, storage, mint, maxt, step, lookbackDelta)
+	// Pre-process the expression to check whether the
+	// expression is step invariant.
+	expr = promql.PreprocessExpr(expr, mint, maxt)
```
Comment (fpetkovski) [inline physicalplan/plan.go:44]: I wonder if we should add this to the logical plan after https://github.com/thanos-community/promql-engine/pull/79?

Comment (yeya24) [reply physicalplan/plan.go:44]: Yeah that's a great idea.
If we have logical plan then dealing with this should be much more easier.

**Comment by yeya24** [comment]
Thanks!

**Pull request merged**